src/backend/distributed/utils/background_jobs.c

```diff
@@ -850,12 +1146,15 @@ UpdateDependingTasks(BackgroundTask *task)
 
 
 /*
- * ConsumeTaskWorkerOutput consumes the output of an executor and mutates the
- * BackgroundTask object to reflect changes like the message and status on the task.
+ * ConsumeTaskWorkerOutput consumes the output of an executor and sets the message as
+ * the last message read from the queue. It also sets hadError as true if executor had
+ * error.
  */
-static void
+static shm_mq_result
 ConsumeTaskWorkerOutput(shm_mq_handle *responseq, StringInfo message, bool *hadError)
```

Comment:
To simplify calling this maybe we should return our own enum with three variants:
1. STILL_RUNNING/WOULDBLOCK/{some better name}
2. SUCCESS
3. ERROR

Then the hadError argument is not needed.